clickhouse/columns/map.h

```diff
@@ -214,13 +221,17 @@ class ColumnMapT : public ColumnMap {
 
     template <typename T>
     inline void Append(const T& value) {
-        // TODO  Refuse to copy.
-        std::vector<std::tuple<typename T::key_type, typename T::mapped_type>> container;
-        container.reserve(value.size());
-        for (const auto& i : value) {
-            container.emplace_back(i.first, i.second);
-        }
-        typed_data_->Append(container.begin(), container.end());
+        using BaseIter = typename T::const_iterator;
+        using KeyOfT = decltype(std::declval<BaseIter>()->first);
+        using ValOfT = decltype(std::declval<BaseIter>()->second);
+        using Functor = std::function<std::tuple<KeyOfT, ValOfT>(const BaseIter&)>;
+        using Iterator = ProjectedIterator<Functor, BaseIter>;
+
+        Functor functor = [](const BaseIter& i) {
+            return std::make_tuple(i->first, i->second);
```

Comment:
Just a random idea: maybe use `std::ref`/`std::cref` here? Because otherwise you are still copying values, not in bulk, but one-by-one.

```diff
-  return std::make_tuple(i->first, i->second);
+  return std::make_tuple(std::cref(i->first), std::cref(i->second));
```